### ISSUE TITLE
False positive reportArgumentType error on list unions

### ISSUE THREAD

**Issue by Weidav** [DetachHead/basedpyright#534]
I'm not 100% sure if this is a bug, but I'm getting this error message: `list[str]" is incompatible with "list[str | DwdObservationDataset | DwdObservationParameter]` from this piece of code. From my understandig this this types should be compatible. Possible this issue is related to https://github.com/DetachHead/basedpyright/issues/122

```python
parameters_10min: = [
    "wind_speed",
    "wind_direction",
    "wind_gust_max",
    "temperature_air_mean_200",
    "pressure_air_site",
]

data_10min = DwdObservationRequest(
    parameter=parameters_10min,
    resolution=Resolution.MINUTE_10,
    period=Period.NOW,
    start_date=start_time,
    end_date=datetime.now(timezone.utc),
)
```

hers the declaration for DwdObservationRequest:
```python
    def __init__(
        self,
        parameter: str
        | DwdObservationDataset
        | DwdObservationParameter
        | list[str | DwdObservationDataset | DwdObservationParameter],
        resolution: str | Resolution | DwdObservationResolution,
        period: str | Period | DwdObservationPeriod | list[str | Period | DwdObservationPeriod] = None,
        start_date: str | dt.datetime | None = None,
        end_date: str | dt.datetime | None = None,
        settings: Settings | None = None,
    ):
```
I can work around this by adding this type hint to my variable: `parameters_10min: list[str | DwdObservationDataset | DwdObservationParameter]`

Here is the full error message:
```

Argument of type "list[str]" cannot be assigned to parameter "parameter" of type "str | DwdObservationDataset | DwdObservationParameter | list[str | DwdObservationDataset | DwdObservationParameter]" in function "__init__"
  Type "list[str]" is incompatible with type "str | DwdObservationDataset | DwdObservationParameter | list[str | DwdObservationDataset | DwdObservationParameter]"
    "list[str]" is incompatible with "str"
    "list[str]" is incompatible with "DwdObservationDataset"
    "list[str]" is incompatible with "DwdObservationParameter"
    "list[str]" is incompatible with "list[str | DwdObservationDataset | DwdObservationParameter]"
      Type parameter "_T@list" is invariant, but "str" is not the same as "str | DwdObservationDataset | DwdObservationParameter"
      Consider switching from "list" to "Sequence" which is covariant
```

**Comment by Weidav** [comment]
I didn't want to include this in the issue, but I need to **thank you so much** for bringing more then 3 painfull years of python development in VSCodium with pyright, then mypy and mainly without the pylance features to an end! 🙏

**Comment by KotlinIsland** [comment]
`list` is invariant on its type parameter. Your error message here is correct because:
```py

def foo(data: list[str | int]):
    data.append(1)


a = ["hello"]
foo(a)
print(a[1] + "!")
```
